### PR TITLE
Ensure kubelet pid is not moved to system container

### LINKED ISSUE #19340: AWS broken by recent commit (?)
When doing kube-up with head on AWS, the containers are not starting on the master.

From journalctl -fu docker

```
Jan 06 19:27:03 ip-172-20-0-9 docker[10807]: time="2016-01-06T19:27:03.523553387Z" level=error msg="Handler for POST /containers/{name:.*}/start returned error: Cannot start container c5d74eedbc37e38f11dcfe97c33c171ec381c044b86cf6d45bf1f6c85c4e0563: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-c5d74eedbc37e38f11dcfe97c33c171ec381c044b86cf6d45bf1f6c85c4e0563.scope/cpu.shares: no such file or directory"
Jan 06 19:27:03 ip-172-20-0-9 docker[10807]: time="2016-01-06T19:27:03.523877582Z" level=error msg="HTTP Error" err="Cannot start container c5d74eedbc37e38f11dcfe97c33c171ec381c044b86cf6d45bf1f6c85c4e0563: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-c5d74eedbc37e38f11dcfe97c33c171ec381c044b86cf6d45bf1f6c85c4e0563.scope/cpu.shares: no such file or directory" statusCode=404
```

From journalctl -fu kubelet

```
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: E0106 19:28:57.275761   10472 manager.go:1811] Failed to create pod infra container: API error (404): Cannot start container fcf82532d1d66ec90cf87c918bb93898e6006bc2ad136e336c4ad9d6c2e4eefe: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-fcf82532d1d66ec90cf87c918bb93898e6006bc2ad136e336c4ad9d6c2e4eefe.scope/cpu.shares: no such file or directory
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: ; Skipping pod "etcd-server-ip-172-20-0-9.ec2.internal_kube-system(89bcb3ce2e0d539d8532f98175510b71)"
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: E0106 19:28:57.370632   10472 manager.go:1811] Failed to create pod infra container: API error (404): Cannot start container abd1c98c30e075e332a38d6087a6dffa245d435ab509887b798e514b1cc90d51: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-abd1c98c30e075e332a38d6087a6dffa245d435ab509887b798e514b1cc90d51.scope/cpu.shares: no such file or directory
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: ; Skipping pod "kube-scheduler-ip-172-20-0-9.ec2.internal_kube-system(1df2bb4d26f706e5720737613e41eb34)"
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: W0106 19:28:57.405391   10472 container.go:329] Failed to create summary reader for "/system/system.slice/docker-fcf82532d1d66ec90cf87c918bb93898e6006bc2ad136e336c4ad9d6c2e4eefe.scope": none of the resources are being tracked.
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: W0106 19:28:57.432928   10472 container.go:329] Failed to create summary reader for "/system/system.slice/docker-abd1c98c30e075e332a38d6087a6dffa245d435ab509887b798e514b1cc90d51.scope": none of the resources are being tracked.
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: E0106 19:28:57.470683   10472 pod_workers.go:125] Error syncing pod 89bcb3ce2e0d539d8532f98175510b71, skipping: API error (404): Cannot start container fcf82532d1d66ec90cf87c918bb93898e6006bc2ad136e336c4ad9d6c2e4eefe: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-fcf82532d1d66ec90cf87c918bb93898e6006bc2ad136e336c4ad9d6c2e4eefe.scope/cpu.shares: no such file or directory
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: E0106 19:28:57.471763   10472 pod_workers.go:125] Error syncing pod 1df2bb4d26f706e5720737613e41eb34, skipping: API error (404): Cannot start container abd1c98c30e075e332a38d6087a6dffa245d435ab509887b798e514b1cc90d51: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-abd1c98c30e075e332a38d6087a6dffa245d435ab509887b798e514b1cc90d51.scope/cpu.shares: no such file or directory
Jan 06 19:28:57 ip-172-20-0-9 kubelet[10472]: E0106 19:28:57.472410   10472 pod_workers.go:125] Error syncing pod 4b1f1bff71eb6da5ccda68bb9165b1b8, skipping: API error (404): Cannot start container 3d9e849e34c42f666319bdca763335f0b31981e653aa45be9b77d9f410b64769: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-3d9e849e34c42f666319bdca763335f0b31981e653aa45be9b77d9f410b64769.scope/cpu.shares: no such file or directory
Jan 06 19:29:00 ip-172-20-0-9 kubelet[10472]: E0106 19:29:00.467485   10472 runtime.go:188] invalid container ID: ""
Jan 06 19:29:00 ip-172-20-0-9 kubelet[10472]: E0106 19:29:00.467549   10472 runtime.go:188] invalid container ID: ""
Jan 06 19:29:00 ip-172-20-0-9 kubelet[10472]: E0106 19:29:00.467598   10472 runtime.go:188] invalid container ID: ""
Jan 06 19:29:00 ip-172-20-0-9 kubelet[10472]: E0106 19:29:00.467634   10472 runtime.go:188] invalid container ID: ""
```

The `docker-<id>` files seem to be in `/sys/fs/cgroup/cpu,cpuacct/system.slice`, not `/sys/fs/cgroup/cpu,cpuacct/system/system.slice`.

I suspect this is because I set the `--system-container` flag without knowing what it really does:

```
> ps -ef | grep docker
root     10472     1  6 19:24 ?        00:00:24 /usr/local/bin/kubelet $DAEMON_ARGS --enable-debugging-handlers=false --cloud-provider=aws --config=/etc/kubernetes/manifests --allow-privileged=True --v=2 --cluster-dns=10.0.0.10 --cluster-domain=cluster.local --docker-root=/mnt/ephemeral/docker --root-dir=/mnt/ephemeral/kubernetes/kubelet --configure-cbr0=true --cgroup-root=docker --system-container=/system --pod-cidr=10.246.0.0/24
root     10807     1  3 19:24 ?        00:00:12 /usr/bin/docker -d -H fd:// -s aufs -g /mnt/ephemeral/docker --bridge=cbr0 --iptables=false --ip-masq=false --log-level=warn
```

Can someone that understands the `--system-container` flag possibly give me a hint as to what the right value is or how I can figure it out?  Or maybe some hint as to why this broke recently?

### LINKED ISSUE #19346: kubelet not running in /kubelet
`/sys/fs/cgroup/cpuacct/kubelet` shows no increase of cpu usage. Jenkins builds also show that the resource monitoring test has been measuring 0 cpu usage for /kubelet as far back as 12/31/2015. The builds before that are no longer available.  

I confirmed the regression in my local cluster, which showed that kubelet is running in /system.

Bisecting using my local cluster, the offending PR seems to be #17883, which bumped cadvisor and migrated from docker/libcontainer to opencontainers/runc/libcontainer).

/cc @kubernetes/goog-node, @jimmidyson

### PR DESCRIPTION
**Pull request by jimmidyson**
Fixes #19346.

/cc @yujuhong @justinsb

### CODE REVIEW

**Comment by jimmidyson** [comment]
Still validating this fix actually...

**Comment by jimmidyson** [comment]
Note that the `cgroup-root` flag breaks running containers on systemd systems as it sets the cgroup parent - docker/docker#17126.

**Comment by yujuhong** [comment]
Verified the fix on a debian machine:
```console
$ ps -o cgroup $(pidof kubelet)
CGROUP
10:net_prio:/kubelet,9:perf_event:/kubelet,8:blkio:/kubelet,7:net_cls:/kubelet,6:freezer:/kubelet,5:devices:/kubelet,4:memory:/kubelet,3:cpuacct:/kubelet,2:cpu:/kubelet,1:cpuset:/kubelet
```

/cc @dchen1107 @vishh to review the fix

**Comment by yujuhong** [comment]
@jimmidyson thanks for the fix! Do you have any idea why this suddenly surfaced?

**Comment by jimmidyson** [comment]
One side effect of this is that pid 1 is moved to `/system`. Not 100% sure that's right but can't see any problems with it. Any thoughts @dchen1107?

**Comment by jimmidyson** [comment]
@yujuhong Been looking this morning. It's to do with the different implementation of `GetPids` in the previous `docker/libcontainer` & the new `opencontainers/runc/libcontainer. Previously (https://github.com/docker/libcontainer/blob/83a102cc68a09d890cce3b6c2e5c14c49e6373a0/cgroups/fs/apply_raw.go#L205-L217) this was not recursive. New implementation (https://github.com/opencontainers/runc/blob/334b93583356c9cb55944f50735ba68fbf120d78/libcontainer/cgroups/utils.go#L326-L347) is recursive.

**Comment by dchen1107** [comment]
@jimmidyson I couldn't think about any concern by moving pid 1 to /system cgroup for now. /system is unlimited today anyway. We do need to protect pid 1 though.

**Comment by dchen1107** [comment]
Refreshed my memory a little bit here: The particular reason we left pid 1 at root cgroup is that on systemd node, libcontainer gets subsystem path through GetInitCgroupDir, and the same code is applying here too:

https://github.com/opencontainers/runc/blob/d97d5e8b007e4657316eed76ea30bc0f690230cf/libcontainer/cgroups/systemd/apply_systemd.go#L403

**Comment by dchen1107** [comment]
cc/ @derekwaynecarr

**Comment by jimmidyson** [comment]
Ah yes - https://github.com/opencontainers/runc/blob/334b93583356c9cb55944f50735ba68fbf120d78/libcontainer/cgroups/utils.go#L186-L194 - still dependent on pid 1. Will prevent it from being moved.

File: pkg/kubelet/cm/container_manager_linux.go [hunk 1]
```diff
@@ -377,6 +382,20 @@ func ensureSystemContainer(rootContainer *fs.Manager, manager *fs.Manager) error
...
+		// PIDs to move.
+		kubeletPids, err := kubeletContainer.GetPids()
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to list PIDs for kubelet: %v", err))
+			continue
+		}
+
+		kubeletCgroupPIDMap := make(map[int]struct{}, len(kubeletPids))
+		for _, pid := range kubeletPids {
+			fmt.Println(pid)
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/container_manager_linux.go]: remove debug code

Comment (jimmidyson) [inline pkg/kubelet/cm/container_manager_linux.go]: Done.

File: pkg/kubelet/cm/container_manager_linux.go [hunk 2]
```diff
@@ -377,6 +382,14 @@ func ensureSystemContainer(rootContainer *fs.Manager, manager *fs.Manager) error
 			continue
 		}
 
+		// Get PIDs in kubelet group so we can remove them from the list of
+		// PIDs to move.
+		kubeletPids, err := kubeletContainer.GetPids()
```
Comment (vishh) [inline pkg/kubelet/cm/container_manager_linux.go]: What about docker container? Won't that also be included if we were to do a recursive walk from `/`?

**Comment by vishh** [comment]
@jimmidyson: Instead of relying on libcontainer here, can we have our own version of GetPids()? 
Another option is to add support in libcontainer to avoid doing recursive walks.

**Comment by justinsb** [comment]
Cherry-picking this onto master did fix the problems with AWS (#19340) :-)

**Comment by jimmidyson** [comment]
@justinsb Interesting - didn't think it would fix it but good :)

**Comment by jimmidyson** [comment]
@vishh Discussing non-recursive `GetPids` at opencontainers/runc#457.

**Comment by jimmidyson** [comment]
Non-recursive `GetPIDs` is merged in libcontainer so waiting for a tagged release of that before updating cadvisor & subsequently this PR.

**Comment by vishh** [comment]
Great. Thanks!

**Comment by brendandburns** [comment]
@k8s-bot test this please

File: pkg/kubelet/cm/container_manager_linux.go (hunk 2 above)
Comment (jimmidyson) [inline pkg/kubelet/cm/container_manager_linux.go]: This is sorted now the `libcontainer.GetPids` is non-recursive.

**Comment by jimmidyson** [comment]
This is waiting on a new tag being applied to `opencontainers/runc` for a necessary libcontainer upgrade. Relevant PR  for tracking is opencontainers/runc#470.

**Comment by yujuhong** [comment]
> This is waiting on a new tag being applied to opencontainers/runc for a necessary libcontainer upgrade. Relevant PR for tracking is opencontainers/runc#470 as this is the last thing before tag is applied AFAIK.

Ping! Seems like the opencontainers issue has been closed/fixed.

**Comment by jimmidyson** [comment]
@yujuhong Slave driver! There were a few more issues that needed to be fixed too which now have been, runc 0.0.7 tag applied, PR in to cadvisor for upgrade then will update this PR with relevant godep upgrades.

**Comment by yujuhong** [comment]
@jimmidyson thanks for the update! Just wanted to get this in sooner to avoid more cherry-picking if possible ;)

**Comment by jimmidyson** [comment]
@yujuhong Agreed

**Comment by jimmidyson** [comment]
@yujuhong Dependencies bumped as required - just want this to turn green now :)

**Comment by jimmidyson** [comment]
I can't see from the logs what the test failure is - cluster won't come up. Works locally for me with `local-up-cluster.sh`.

**Comment by jimmidyson** [comment]
Bringing an e2e cluster up with vagrant works fine too...

Can someone try to rekick this please?

**Comment by vishh** [comment]
@k8s-bot: ok to test

**Comment by jimmidyson** [comment]
Can someone help me identify why the cluster isn't starting up on GCE while it is on Vagrant for me locally? There is nothing in the logs that helps me so would need someone with access to the GCE nodes I assume.

**Comment by jimmidyson** [comment]
I've just spun up a cluster on GCE via `go run hack/e2e.go -v --up` from this PR branch & cluster starts up properly...

**Comment by jimmidyson** [comment]
Think this is working now - waiting on e2e tests.

**Comment by jimmidyson** [comment]
Don't know why e2e tests aren't even starting... Anyone care to kick it?

**Comment by justinsb** [comment]
@k8s-bot test this

**Comment by jimmidyson** [comment]
This looks like a general problem with the e2e & integration test builders. Must have patience :)

**Comment by jimmidyson** [comment]
Rebased.

**Comment by jimmidyson** [comment]
Yay e2e is green!

**Comment by jimmidyson** [comment]
@vishh @yujuhong This required a rebase so ok-to-merge label has been removed. All green again though so can someone re-add the label? Would like to get this merged.

**Comment by dchen1107** [comment]
LGTM

It is ok for now, but could we please make sure next time we separate gedep updates from the real core changes. Thanks!

**Comment by jimmidyson** [comment]
When you say separate you mean in separate commits? Would have to be in same PR as wouldn't compile otherwise.

**Comment by ncdc** [comment]
@jimmidyson changes to Godeps in 1 commit, changes to everything else in 1 or more separate commits. Same PR.

**Comment by justinsb** [comment]
An earlier version of this PR _did_ fix the problems on AWS, but the version that we merged did not.

@jimmidyson any idea what changed between 3 weeks ago and now with this PR?

Docker is once again reporting:

```
Jan 31 18:03:22 ip-172-20-0-9 docker[11233]: time="2016-01-31T18:03:22.784983514Z" level=error msg="Handler for POST /containers/{name:.*}/start returned error: Cannot start container 4295b84943fbfa285ef835937411ee6b114422b2f0d6d81e0e35c74d9979bfa8: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-4295b84943fbfa285ef835937411ee6b114422b2f0d6d81e0e35c74d9979bfa8.scope/cpu.shares: no such file or directory"
Jan 31 18:03:22 ip-172-20-0-9 docker[11233]: time="2016-01-31T18:03:22.785485882Z" level=error msg="HTTP Error" err="Cannot start container 4295b84943fbfa285ef835937411ee6b114422b2f0d6d81e0e35c74d9979bfa8: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-4295b84943fbfa285ef835937411ee6b114422b2f0d6d81e0e35c74d9979bfa8.scope/cpu.shares: no such file or directory" statusCode=404
Jan 31 18:03:22 ip-172-20-0-9 docker[11233]: time="2016-01-31T18:03:22.799153179Z" level=error msg="Handler for POST /containers/{name:.*}/start returned error: Cannot start container 849c7fcd59d515e04eb50424fcad036502cfbfe862150bc81db60a210a0c853d: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-849c7fcd59d515e04eb50424fcad036502cfbfe862150bc81db60a210a0c853d.scope/cpu.shares: no such file or directory"
Jan 31 18:03:22 ip-172-20-0-9 docker[11233]: time="2016-01-31T18:03:22.799190315Z" level=error msg="HTTP Error" err="Cannot start container 849c7fcd59d515e04eb50424fcad036502cfbfe862150bc81db60a210a0c853d: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system/system.slice/docker-849c7fcd59d515e04eb50424fcad036502cfbfe862150bc81db60a210a0c853d.scope/cpu.shares: no such file or directory" statusCode=404
```

**Comment by jimmidyson** [comment]
Quite a bit changed tbh, mostly simplifications. I was surprised when this originally fixed your issue. Shall we reopen your original issue & discuss there? I assume it has something to do with docker running in a non root cgroup. Could you share output of `ps -o cgroup $(pidof docker)` (or just replace docker pid in there)?

**Comment by ncdc** [comment]
Does the file not found error occur 100% of the time?

**Comment by justinsb** [comment]
@jimmidyson output of `ps -o cgroup $(pidof docker)`

```
CGROUP
10:freezer:/system.slice,9:perf_event:/system.slice,8:blkio:/system.slice/docker.service,7:devices:/system.slice/docker.service,6:memory:/system.slice/docker.service,5:hugetl
```

@ncdc the error does appear to happen 100% of the time, yes.  There are no containers listed in `docker ps`, there are a lot of pause containers listed in `docker ps -a` (output below) and AFAICT each docker container immediately gives the file-not-found error I pasted before.

```
> docker ps -a
ONTAINER ID        IMAGE                                COMMAND             CREATED              STATUS              PORTS               NAMES
5b37b21b9a6c        gcr.io/google_containers/pause:2.0   "/pause"            4 seconds ago        Created                                 k8s_POD.6059dfa2_kube-scheduler-ip-1
72-20-0-9.ec2.internal_kube-system_c9bc5a66e6d1ac4e107961af3ad2bcca_fcf872d4
74d3c81f6853        gcr.io/google_containers/pause:2.0   "/pause"            5 seconds ago        Created                                 k8s_POD.6059dfa2_etcd-server-ip-172-
20-0-9.ec2.internal_kube-system_89bcb3ce2e0d539d8532f98175510b71_76bbf189
f83e6b351ffc        gcr.io/google_containers/pause:2.0   "/pause"            5 seconds ago        Created                                 k8s_POD.6059dfa2_kube-controller-man
ager-ip-172-20-0-9.ec2.internal_kube-system_c6c484fd6bbec33521b4d9a1341953ca_b856d129
52c0da8397f1        gcr.io/google_containers/pause:2.0   "/pause"            5 seconds ago        Created                                 k8s_POD.6059dfa2_fluentd-elasticsear
ch-ip-172-20-0-9.ec2.internal_kube-system_a386fa51642994629a8b037fd291b2c8_bdd9a5bd
6c79550d8947        gcr.io/google_containers/pause:2.0   "/pause"            6 seconds ago        Created                                 k8s_POD.6059dfa2_kube-apiserver-ip-1
72-20-0-9.ec2.internal_kube-system_0b86d21b158c05f0afa46fdf39853024_93fe6416
...
```

I agree it doesn't really make sense to discuss this here (on a merged PR) - I was just hoping it might be something simple that changed.  I am going to resume discussion in #19340

**Pull request merged**